File: src/test/java/org/example/eventhub/mapper/TicketMapperTest.java
package org.example.eventhub.mapper;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.math.BigDecimal;
import org.example.eventhub.dto.event.EventResponseShort;
import org.example.eventhub.dto.ticket.TicketResponseLong;
import org.example.eventhub.dto.ticket.TicketResponseShort;
import org.example.eventhub.dto.user.UserResponseShort;
import org.example.eventhub.entity.Event;
import org.example.eventhub.entity.Order;
import org.example.eventhub.entity.Ticket;
import org.example.eventhub.entity.User;
import org.example.eventhub.enums.TicketStatus;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class TicketMapperTest {

    @Mock
    private EventMapper eventMapper;

    @Mock
    private UserMapper userMapper;

    @InjectMocks
    private TicketMapper ticketMapper;

    private User user;
    private Event event;
    private Order order;
    private Ticket ticket;

    private UserResponseShort userShort;
    private EventResponseShort eventShort;

    @BeforeEach
    void setUp() {
        user = User.builder().id(10L).username("john_doe").build();

        event = Event.builder().id(100L).title("Concert").build();

        order = Order.builder().id(1L).build();

        ticket = Ticket.builder()
                .id(1000L)
                .event(event)
                .user(user)
                .order(order)
                .price(BigDecimal.valueOf(2500))
                .status(TicketStatus.RESERVED)
                .build();

        userShort = new UserResponseShort("john_doe");
        eventShort = new EventResponseShort(100L, "Concert", null);
        when(userMapper.toShortDto(user)).thenReturn(userShort);
        when(eventMapper.toShortDto(event)).thenReturn(eventShort);
    }

    @Test
    @DisplayName("toLongDto: возвращает правильный Long DTO со всеми полями")
    void toLongDto_shouldReturnCorrectDto() {
        TicketResponseLong expected = new TicketResponseLong(
                1000L, eventShort, userShort, 1L, BigDecimal.valueOf(2500), TicketStatus.RESERVED);

        assertEquals(expected, ticketMapper.toLongDto(ticket));
    }

    @Test
    @DisplayName("toShortDto: возвращает правильный Short DTO без id, orderId и price")
    void toShortDto_shouldReturnCorrectDto() {
        TicketResponseShort expected = new TicketResponseShort(eventShort, userShort, TicketStatus.RESERVED);

        assertEquals(expected, ticketMapper.toShortDto(ticket));
    }

    @Test
    @DisplayName("toLongDto: корректно маппит вложенные объекты через зависимости")
    void toLongDto_mapsNestedObjectsCorrectly() {
        ticketMapper.toLongDto(ticket);

        verify(eventMapper).toShortDto(event);
        verify(userMapper).toShortDto(user);
    }

    @Test
    @DisplayName("toShortDto: корректно маппит вложенные объекты через зависимости")
    void toShortDto_mapsNestedObjectsCorrectly() {
        ticketMapper.toShortDto(ticket);

        verify(eventMapper).toShortDto(event);
        verify(userMapper).toShortDto(user);
    }
}

File: src/test/kotlin/org/example/eventhub/mapper/TicketMapperTest.kt
package org.example.eventhub.mapper

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import org.example.eventhub.dto.event.EventResponseShort
import org.example.eventhub.dto.user.UserResponseShort
import org.example.eventhub.entity.*
import org.example.eventhub.enums.*
import java.math.BigDecimal
import java.time.LocalDateTime

class TicketMapperTest : StringSpec({

    val eventMapper = mockk<EventMapper>()
    val userMapper = mockk<UserMapper>()

    val mapper = TicketMapper(eventMapper, userMapper)

    val userDto = UserResponseShort("user")
    val eventDto = EventResponseShort(
        id = 1L,
        title = "Event",
        dateTime = LocalDateTime.of(2021, 1, 1, 0, 0)
    )

    lateinit var user: User
    lateinit var event: Event
    lateinit var order: Order
    lateinit var ticket: Ticket

    beforeEach {
        user = User(
            id = 1L,
            username = "user",
            email = "mail@mail.com",
            password = "pass"
        )

        event = Event(
            id = 1L,
            title = "Event",
            dateTime = LocalDateTime.of(2021, 1, 1, 0, 0),
            capacity = 10,
            price = BigDecimal.TEN,
            organizer = user,
            eventStatus = EventStatus.PUBLISHED
        )

        order = Order(
            id = 100L,
            user = user,
            tickets = mutableListOf(),
            totalPrice = BigDecimal.TEN,
            status = OrderStatus.PAID,
            reservedUntil = LocalDateTime.now().plusMinutes(15)
        )

        ticket = Ticket(
            id = 5L,
            event = event,
            user = user,
            order = order,
            price = BigDecimal.TEN,
            status = TicketStatus.RESERVED
        )
    }

    "toShortDto маппит ticket в short dto" {
        every { eventMapper.toShortDto(event) } returns eventDto
        every { userMapper.toShortDto(user) } returns userDto

        val dto = mapper.toShortDto(ticket)

        dto.event shouldBe eventDto
        dto.user shouldBe userDto
        dto.ticketStatus shouldBe TicketStatus.RESERVED
    }

    "toLongDto маппит ticket в long dto" {
        every { eventMapper.toShortDto(event) } returns eventDto
        every { userMapper.toShortDto(user) } returns userDto

        val dto = mapper.toLongDto(ticket)

        dto.id shouldBe 5L
        dto.event shouldBe eventDto
        dto.user shouldBe userDto
        dto.orderId shouldBe 100L
        dto.price shouldBe BigDecimal.TEN
        dto.ticketStatus shouldBe TicketStatus.RESERVED
    }
})
